src/internal/types.ts

```diff
@@ -276,3 +276,99 @@ export type ValueFromNotification<T> = T extends { kind: 'N' | 'E' | 'C' }
 export type Falsy = null | undefined | false | 0 | -0 | 0n | '';
 
 export type TruthyTypesOf<T> = T extends Falsy ? never : T;
+
+/**
+ * One of those magic types that can extract types from an
+ * overloaded function. See https://github.com/ReactiveX/rxjs/issues/5942
+ * for more details
+ */
+type OverloadedArgumentsAndReturnType<T> = T extends {
```

Comment:
I don't understand this at all. haha. T extends a function that.. looks the same in 10 cases? But... isn't the same? But it's the same... but different.

![same-but-different](https://user-images.githubusercontent.com/1540597/109747117-3046e800-7b9c-11eb-86ed-8033aa6d2cf9.gif)

Reply:
Wait... this is a type that is a "better `Parameters<T>`"? I still don't quite understand how it works.

Reply:
Yeah this is the magical overloaded function inference technique described at https://stackoverflow.com/questions/52760509/typescript-returntype-of-overloaded-function/52761156#52761156